### PR TITLE
Make TypeDelegate ctor/create accept a TypeFunction

### PR DESCRIPTION
**Pull request by Geod24**
```
A TypeFunction is the only legal type that can be passed as argument
to TypeDelegate constructor, however the constructor wasn't aware of this.
```

### CODE REVIEW

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -4113,7 +4113,7 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         // Type is a "delegate to" or "pointer to" the function literal
         if ((exp.fd.isNested() && exp.fd.tok == TOK.delegate_) || (exp.tok == TOK.reserved && exp.fd.treq && exp.fd.treq.ty == Tdelegate))
         {
-            exp.type = new TypeDelegate(exp.fd.type);
+            exp.type = new TypeDelegate(cast(TypeFunction)exp.fd.type);
```
Comment (ibuclaw) [inline src/dmd/expressionsem.d:4116]: toTypeFunction?

Comment (Geod24) [reply src/dmd/expressionsem.d:4116]: Doesn't make a difference, does it ? `fd.type` should really be typed as `TypeFunction`.

Comment (ibuclaw) [reply src/dmd/expressionsem.d:4116]: Nothing wrong with a bit of enforcement then, and an assert will be thrown if someone accidentally assigns something else.

Comment (Geod24) [reply src/dmd/expressionsem.d:4116]: Done

Comment (ibuclaw) [reply src/dmd/expressionsem.d]: There are other casts.

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -4113,7 +4113,7 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         // Type is a "delegate to" or "pointer to" the function literal
         if ((exp.fd.isNested() && exp.fd.tok == TOK.delegate_) || (exp.tok == TOK.reserved && exp.fd.treq && exp.fd.treq.ty == Tdelegate))
         {
-            exp.type = new TypeDelegate(exp.fd.type);
+            exp.type = new TypeDelegate(exp.fd.type.toTypeFunction());
```
Comment (WalterBright) [inline src/dmd/expressionsem.d:4116]: change toTypeFunction to isTypeFunction. The latter will return null if it is not, and the next line will fault. Everything else uses the `is` functions, we shouldn't be created a same-only-different parallel set.

File: src/dmd/expressionsem.d [hunk 3]
```diff
@@ -6588,7 +6588,7 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
 
         e.e1 = e.e1.expressionSemantic(sc);
 
-        e.type = new TypeDelegate(e.func.type);
+        e.type = new TypeDelegate(cast(TypeFunction) e.func.type);
```
Comment (WalterBright) [inline src/dmd/expressionsem.d:6591]: `e.func.type.isTypeFunction()`

File: src/dmd/mtype.d [hunk 4]
```diff
@@ -2332,7 +2332,7 @@ extern (C++) abstract class Type : ASTNode
                 }
                 else if (ty == Tdelegate)
                 {
-                    t = new TypeDelegate(t);
+                    t = new TypeDelegate(cast(TypeFunction) t);
```
Comment (WalterBright) [inline src/dmd/mtype.d:2335]: `t.isTypeFunction()`

File: src/dmd/mtype.d [hunk 5]
```diff
@@ -5288,11 +5288,11 @@ extern (C++) final class TypeDelegate : TypeNext
 
     override TypeDelegate syntaxCopy()
     {
-        Type t = next.syntaxCopy();
-        if (t == next)
+        auto tf = cast(TypeFunction) next.syntaxCopy();
```
Comment (WalterBright) [inline src/dmd/mtype.d:5291]: `next.syntaxCopy().isTypeFunction()`

**Review by WalterBright — DISMISSED** [review]
replace all casts with `is` functions, which are type-safe

File: src/dmd/target.d [hunk 6]
```diff
@@ -1379,8 +1379,8 @@ struct TargetCPP
         else if (p.storageClass & STC.lazy_)
         {
             // Mangle as delegate
-            Type td = new TypeFunction(ParameterList(), t, LINK.d);
-            td = new TypeDelegate(td);
+            auto tf = new TypeFunction(ParameterList(), t, LINK.d);
+            auto td = new TypeDelegate(tf);
             t = merge(t);
```
Comment (Geod24) [inline src/dmd/target.d:1384]: @WalterBright : This looks like a bug to me. Shouldn't it be `t = merge(td);` ?

**Comment by Geod24** [comment]
Review applied

**Review by RazvanN7 — APPROVED** [review]

**Review by thewilsonator — APPROVED** [review]

**Comment by quickfur** [comment]
This PR introduced an ICE: https://issues.dlang.org/show_bug.cgi?id=22686

**Pull request merged**